### PR TITLE
feat(backend): add filterState parameter to listAllContributions query

### LINKED ISSUE #2594: 🚀 [Feature] [Admin] Contributions List
## 🚀 Feature

[Admin] Contributions List

Currently you can only access
- contributions per user
- open contributions

### TODOs
- [ ] Make the open contributions page always visible in the menu.
- [ ] The page must contain two tabs (open contributions, all contributions)
- [ ] if the open contributions contains items a red badge with the number of open contributions must be shown in the menu and on the tab, aswell as on the overview page.
- [ ] the `all contributions` must be ordered by the contribution with the last action on it (`updated_at`?) - e.g. confirmed, denied, deleted, ...
- [ ] additional tabs for filters are thinkable: `deleted`, `denied`, `confirmed`. All of those must be queried by the same graphQL query taking a list of filters(e.g. no filter = all, [' denied', 'confirmed'] for those two)

### PR DESCRIPTION
**Pull request by Elweyn**
## 🍰 Pullrequest

Add a string or null parameter to the listAllContributions query, so that we can filter on statis ('DELETED', 'DENIED', ..., 'CONFIRMED')

### Issues

- fixes #2594 

### Todo

- [X] Add filterState array of ContributionStatus

### CODE REVIEW

File: backend/src/graphql/resolver/ContributionResolver.ts [hunk 1]
```diff
@@ -179,12 +180,47 @@ export class ContributionResolver {
...
+          logger.error(
+            `${i18n.__('error.contributions.wrongFilterState', {
+              contributionState: filterStates[i],
+            })}`,
+          )
+          throw new Error(
+            `${i18n.__('error.contributions.wrongFilterState', {
+              contributionState: filterStates[i],
+            })}`,
+          )
```
Comment (Elweyn) [inline backend/src/graphql/resolver/ContributionResolver.ts:209]: Open to discussion if we want it or not! Is only a test of mine to see if I understand the i18n api right!

File: backend/src/graphql/resolver/ContributionResolver.test.ts [hunk 2]
```diff
@@ -681,7 +681,7 @@ describe('ContributionResolver', () => {
               currentPage: 1,
               pageSize: 25,
               order: 'DESC',
-              filterConfirmed: false,
+              filterState: null,
```
Comment (Elweyn) [inline backend/src/graphql/resolver/ContributionResolver.test.ts:684]: Needs more tests!

File: backend/src/graphql/resolver/ContributionResolver.ts [hunk 3]
```diff
@@ -179,12 +180,47 @@ export class ContributionResolver {
   async listAllContributions(
     @Args()
     { currentPage = 1, pageSize = 5, order = Order.DESC }: Paginated,
+    @Arg('filterState', () => [ContributionStatus], { nullable: true })
```
Comment (Mogge) [inline backend/src/graphql/resolver/ContributionResolver.ts:183]: On the left hand side you have State and the right hand side is Status. I prefer Status.

**Review by Mogge — CHANGES_REQUESTED** [review]
Please use `status` instead of `state`

File: backend/src/graphql/resolver/ContributionResolver.ts (hunk 3 above)
Comment (Elweyn) [reply backend/src/graphql/resolver/ContributionResolver.ts]: Change everything to statusFilter

File: backend/src/locales/en.json [hunk 4]
```diff
@@ -1,4 +1,5 @@
 {
+  "contributions": {},
```
Comment (Mogge) [inline backend/src/locales/en.json:2]: Why this?

File: backend/src/graphql/resolver/ContributionResolver.ts [hunk 5]
```diff
@@ -179,12 +179,36 @@ export class ContributionResolver {
...
+    } = {}
+    const typeStatus = Object.values(ContributionStatus)
+    if (statusFilters !== null) {
+      const statusFilterArray = []
+      const length = statusFilters.length
+      let i = 0
+      for (i; i < length; i++) {
+        const statusFilter = statusFilters[i]
+        if (!typeStatus.includes(statusFilter)) {
+          logger.error(`Nicht definierter Status wurde gesendet: ${statusFilters[i]}`)
```
Comment (Mogge) [inline backend/src/graphql/resolver/ContributionResolver.ts:196]: Can this even happen? Does the schema throw?

File: backend/src/graphql/resolver/ContributionResolver.ts [hunk 6]
```diff
@@ -179,12 +179,36 @@ export class ContributionResolver {
...
     @Args()
     { currentPage = 1, pageSize = 5, order = Order.DESC }: Paginated,
+    @Arg('statusFilter', () => [ContributionStatus], { nullable: true })
+    statusFilters: ContributionStatus[] | null,
   ): Promise<ContributionListResult> {
+    const where: {
+      contributionStatus?: FindOperator<string> | null
+    } = {}
+    const typeStatus = Object.values(ContributionStatus)
+    if (statusFilters !== null) {
```
Comment (Mogge) [inline backend/src/graphql/resolver/ContributionResolver.ts:189]: can be undefined. Therefore:
```suggestion
    if (statusFilters && statusFilter.length) {
```

File: backend/src/graphql/resolver/ContributionResolver.ts [hunk 7]
```diff
@@ -179,12 +179,36 @@ export class ContributionResolver {
...
+      const statusFilterArray = []
+      const length = statusFilters.length
+      let i = 0
+      for (i; i < length; i++) {
```
Comment (Mogge) [inline backend/src/graphql/resolver/ContributionResolver.ts:193]: What about:
```
 where.contributionStatus = In(statusFilters)

File: backend/src/graphql/resolver/ContributionResolver.ts (hunk 5 above)
Comment (Mogge) [reply backend/src/graphql/resolver/ContributionResolver.ts:196]: `"Enum \"ContributionStatus\" cannot represent non-enum value: \"NOT_VALID\".",`

Comment (Mogge) [reply backend/src/graphql/resolver/ContributionResolver.ts:196]: `Value \"PENDIG\" does not exist in \"ContributionStatus\" enum. Did you mean the enum value \"PENDING\"?`

File: backend/src/graphql/resolver/ContributionResolver.ts (hunk 7 above)
Comment (Mogge) [reply backend/src/graphql/resolver/ContributionResolver.ts:193]: works for me

**Review by Mogge — CHANGES_REQUESTED** [review]
Looks very good

File: backend/src/graphql/resolver/ContributionResolver.ts [hunk 8]
```diff
@@ -179,12 +179,23 @@ export class ContributionResolver {
   async listAllContributions(
     @Args()
     { currentPage = 1, pageSize = 5, order = Order.DESC }: Paginated,
+    @Arg('statusFilter', () => [ContributionStatus], { nullable: true })
+    statusFilters: ContributionStatus[] | null,
```
[the remark this answers is not in the document]
Comment (Mogge) [reply backend/src/graphql/resolver/ContributionResolver.ts:183]: Then `undefined` means that there are no filters to apply

File: backend/src/graphql/resolver/ContributionResolver.ts [hunk 9]
```diff
@@ -179,12 +179,23 @@ export class ContributionResolver {
...
   ): Promise<ContributionListResult> {
+    const where: {
+      contributionStatus?: FindOperator<string> | null
+    } = {}
+
+    if (statusFilters !== null && statusFilters.length) {
```
[the remark this answers is not in the document]
Comment (Mogge) [reply backend/src/graphql/resolver/ContributionResolver.ts:189]: In the case of undefined you get an exception

File: backend/src/graphql/resolver/ContributionResolver.test.ts [hunk 10]
```diff
@@ -718,6 +719,28 @@ describe('ContributionResolver', () => {
...
 
+      it('throws an error with non existing filter', async () => {
+        await expect(
+          query({
+            query: listAllContributions,
+            variables: {
+              currentPage: 1,
+              pageSize: 25,
+              order: 'DESC',
+              statusFilter: ['CONFIRMD'],
```
Comment (Mogge) [inline backend/src/graphql/resolver/ContributionResolver.test.ts:730]:
```suggestion
              statusFilter: ['NOT_VALID'],
```
And the exception as expetced

File: backend/src/graphql/resolver/ContributionResolver.ts [hunk 11]
```diff
@@ -179,12 +179,23 @@ export class ContributionResolver {
   async listAllContributions(
     @Args()
     { currentPage = 1, pageSize = 5, order = Order.DESC }: Paginated,
+    @Arg('statusFilter', () => [ContributionStatus], { nullable: true })
```
Comment (Mogge) [inline backend/src/graphql/resolver/ContributionResolver.ts:182]: Is this working?
```suggestion
    @Arg('statusFilter', () => [ContributionStatus!], { nullable: true })
```

**Review by Mogge — CHANGES_REQUESTED** [review]
I still do not understand why you want to force the optional argument to be null

File: backend/src/seeds/graphql/queries.ts [hunk 12]
```diff
@@ -172,18 +172,23 @@ export const listContributions = gql`
 `
 
 export const listAllContributions = `
-query ($currentPage: Int = 1, $pageSize: Int = 5, $order: Order = DESC) {
-  listAllContributions(currentPage: $currentPage, pageSize: $pageSize, order: $order) {
+query ($currentPage: Int = 1, $pageSize: Int = 5, $order: Order = DESC, $statusFilter: [ContributionStatus!] = null) {
```
[the remark this answers is not in the document]
Comment (Elweyn) [reply backend/src/seeds/graphql/queries.ts]: Changed it to the undefined possibility, it is smoother than the previous attempt

File: backend/src/graphql/resolver/ContributionResolver.test.ts [hunk 13]
```diff
@@ -681,7 +682,7 @@ describe('ContributionResolver', () => {
               currentPage: 1,
               pageSize: 25,
               order: 'DESC',
-              filterConfirmed: false,
+              statusFilter: null,
```
[the remark this answers is not in the document]
Comment (Elweyn) [reply backend/src/graphql/resolver/ContributionResolver.test.ts:685]: I would prefer to let it in cause it is the first call to the method, so we can see all the possible parameters here.
I added a test for the case where statusFilter is missing to assure to that we still get the full list.

File: backend/src/graphql/resolver/ContributionResolver.ts (hunk 11 above)
Comment (Elweyn) [reply backend/src/graphql/resolver/ContributionResolver.ts]: Is not working for me, it is writting the following error:
```
Forbidden non-null assertion.eslint[@typescript-eslint/no-non-null-assertion](https://github.com/typescript-eslint/typescript-eslint/blob/v4.33.0/packages/eslint-plugin/docs/rules/no-non-null-assertion.md)
```
I could add a line with
```
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
```
But i am not sure if we want it.

Comment (Mogge) [reply backend/src/graphql/resolver/ContributionResolver.ts:182]: OK. Thanks for trying. Leave it as it is

**Review by Mogge — CHANGES_REQUESTED** [review]
Some small changes for consistency of naming. I suggest a following PR with creations in progress, denied and deleted in the database. The deleted contribution should never be returned

File: backend/src/graphql/resolver/ContributionResolver.test.ts [hunk 14]
```diff
@@ -718,15 +719,371 @@ describe('ContributionResolver', () => {
...
+            errors: [
+              new UserInputError(
+                'Variable "$statusFilter" got invalid value null at "statusFilter[0]"; Expected non-nullable type "ContributionStatus!" not to be null.',
+              ),
+            ],
+          }),
+        )
+      })
+
+      it('throws an error with null and "NOT_VALID" in statusFilter', async () => {
```
[the remark this answers is not in the document]
Comment (Elweyn) [reply backend/src/graphql/resolver/ContributionResolver.test.ts:766]: Check how the method react with multiple error in statusFilter. Could stay as the only test for wrong values in statusFilter

File: backend/src/graphql/resolver/ContributionResolver.ts [hunk 15]
```diff
@@ -179,12 +179,23 @@ export class ContributionResolver {
...
+    statusFilters?: ContributionStatus[],
```
[the remark this answers is not in the document]
Comment (Elweyn) [reply backend/src/graphql/resolver/ContributionResolver.ts:183]: Don't like this it is an array of statusFilter so I prefer statusFilters as variable name.

File: backend/src/graphql/resolver/ContributionResolver.ts [hunk 16]
```diff
@@ -179,12 +179,23 @@ export class ContributionResolver {
...
   ): Promise<ContributionListResult> {
+    const where: {
+      contributionStatus?: FindOperator<string> | null
+    } = {}
+
+    if (statusFilters && statusFilters.length) {
```
[the remark this answers is not in the document]
Comment (Elweyn) [reply backend/src/graphql/resolver/ContributionResolver.ts:189]: Don't like this it is an array of statusFilter so I prefer statusFilters as variable name.

File: backend/src/graphql/resolver/ContributionResolver.ts [hunk 17]
```diff
@@ -179,12 +179,23 @@ export class ContributionResolver {
...
+      where.contributionStatus = In(statusFilters)
```
[the remark this answers is not in the document]
Comment (Elweyn) [reply backend/src/graphql/resolver/ContributionResolver.ts:190]: Don't like this it is an array of statusFilter so I prefer statusFilters as variable name.

**Comment by Elweyn** [comment]
To do: Follow issue 
- [ ] Add IN_PROGRESS contributions in the database
- [ ] Add DENIED contributions in the database
- [ ] Add DELETED contributions in the database

File: backend/src/graphql/resolver/ContributionResolver.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (Mogge) [reply backend/src/graphql/resolver/ContributionResolver.ts:189]: Yes, it is an array, but it is one filter for the status

[the remark this answers is not in the document]
Comment (Mogge) [reply backend/src/graphql/resolver/ContributionResolver.ts:189]: What is plural is statuses, you could say something like `allowedStatuses`

**Review by Mogge — APPROVED** [review]

File: backend/src/graphql/resolver/ContributionResolver.test.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (Elweyn) [reply backend/src/graphql/resolver/ContributionResolver.test.ts:766]: https://github.com/gradido/gradido/issues/2653

**Review by Mogge — APPROVED** [review]

**Pull request merged**